# SPDX-License-Identifier: LGPL-2.1-only OR GPL-3.0-only
# Copyright (C) 2026 OpenKaryon
#
# Windows 9x: font enumeration, creation and metrics through the ANSI GDI
# calls (EnumFontFamiliesExA, CreateFontIndirectA, GetTextMetricsA, ...),
# with the Unicode ones kept on NT. From the GUI removal commit 55137901,
# adapted to 4.8.7.
#
--- a/src/gui/text/qfontengine_win_p.h
+++ b/src/gui/text/qfontengine_win_p.h
@@ -80,7 +80,7 @@
     virtual void addGlyphsToPath(glyph_t *glyphs, QFixedPoint *positions, int nglyphs,
                          QPainterPath *path, QTextItem::RenderFlags flags);
 
-    HGDIOBJ selectDesignFont() const;
+    HGDIOBJ selectDesignFont(QFixed *) const;
 
     virtual glyph_metrics_t boundingBox(const QGlyphLayout &glyphs);
     virtual glyph_metrics_t boundingBox(glyph_t g) { return boundingBox(g, QTransform()); }
@@ -123,12 +123,16 @@
     QString     uniqueFamilyName;
     HFONT       hfont;
     LOGFONT     logfont;
-    uint        stockFont  : 1;
-    uint        ttf        : 1;
+    uint        stockFont   : 1;
+    uint        useTextOutA : 1;
+    uint        ttf         : 1;
     uint        hasOutline : 1;
     uint        cffTable   : 1;
-    TEXTMETRIC  tm;
-    int         lw;
+    union {
+        TEXTMETRICW        w;
+        TEXTMETRICA        a;
+    } tm;
+    int                lw;
     const unsigned char *cmap;
     QByteArray cmapTable;
     mutable qreal lbearing;
--- a/src/gui/text/qfont_win.cpp
+++ b/src/gui/text/qfont_win.cpp
@@ -39,6 +39,11 @@
 **
 ****************************************************************************/
 
+// the miscrosoft platform SDK says that the Unicode versions of
+// TextOut and GetTextExtentsPoint32 are supported on all platforms, so we use them
+// exclusively to simplify code, save a lot of conversions into the local encoding
+// and have generally better unicode support :)
+
 #include "qfont.h"
 #include "qfont_p.h"
 #include "qfontengine_p.h"
@@ -63,7 +68,8 @@
 // ### maybe move to qapplication_win
 QFont qt_LOGFONTtoQFont(LOGFONT& lf, bool /*scale*/)
 {
-    QString family = QString::fromWCharArray(lf.lfFaceName);
+    QString family = QT_WA_INLINE(QString::fromUtf16((ushort*)lf.lfFaceName),
+                                   QString::fromLocal8Bit((char*)lf.lfFaceName));
     QFont qf(family);
     qf.setItalic(lf.lfItalic);
     if (lf.lfWeight != FW_DONTCARE)
--- a/src/gui/text/qfontdatabase_win.cpp
+++ b/src/gui/text/qfontdatabase_win.cpp
@@ -201,12 +201,21 @@
     QString i18n_name;
 
     HDC hdc = GetDC( 0 );
-    LOGFONT lf;
-    memset(&lf, 0, sizeof(LOGFONT));
-    memcpy(lf.lfFaceName, familyName.utf16(), qMin(LF_FACESIZE, familyName.length()) * sizeof(wchar_t));
-    lf.lfCharSet = DEFAULT_CHARSET;
-    HFONT hfont = CreateFontIndirect(&lf);
-
+    HFONT hfont;
+    QT_WA( {
+        LOGFONTW lf;
+        memset( &lf, 0, sizeof( LOGFONTW ) );
+        memcpy( lf.lfFaceName, familyName.utf16(), qMin(LF_FACESIZE, familyName.length())*sizeof(QChar) );
+        lf.lfCharSet = DEFAULT_CHARSET;
+        hfont = CreateFontIndirectW( &lf );
+    }, {
+        LOGFONTA lf;
+        memset( &lf, 0, sizeof( LOGFONTA ) );
+        QByteArray lfam = familyName.toLocal8Bit();
+        memcpy( lf.lfFaceName, lfam, qMin(LF_FACESIZE, lfam.size()) );
+        lf.lfCharSet = DEFAULT_CHARSET;
+        hfont = CreateFontIndirectA( &lf );
+    } );
     if(!hfont) {
         ReleaseDC(0, hdc);
         return QString();
@@ -244,6 +253,32 @@
 
 extern QFont::Weight weightFromInteger(int weight); // qfontdatabase.cpp
 
+static void getFontSignature(const QString &familyName,
+                             NEWTEXTMETRICEX *textmetric,
+                             FONTSIGNATURE *signature)
+{
+    QT_WA({
+        Q_UNUSED(familyName);
+        *signature = textmetric->ntmFontSig;
+    }, {
+        // the textmetric structure we get from EnumFontFamiliesEx on Win9x has
+        // a FONTSIGNATURE, but that one is uninitialized and doesn't work. Have to go
+        // the hard way and load the font to find out.
+        HDC hdc = GetDC(0);
+        LOGFONTA lf;
+        memset(&lf, 0, sizeof(LOGFONTA));
+        QByteArray lfam = familyName.toLocal8Bit();
+        memcpy(lf.lfFaceName, lfam.data(), qMin(LF_FACESIZE, lfam.length()));
+        lf.lfCharSet = DEFAULT_CHARSET;
+        HFONT hfont = CreateFontIndirectA(&lf);
+        HGDIOBJ oldobj = SelectObject(hdc, hfont);
+        GetTextCharsetInfo(hdc, signature, 0);
+        SelectObject(hdc, oldobj);
+        DeleteObject(hfont);
+        ReleaseDC(0, hdc);
+    });
+}
+
 static
 void addFontToDatabase(QString familyName, const QString &scriptName,
                        TEXTMETRIC *textmetric,
@@ -261,17 +296,26 @@
     bool scalable;
     int size;
 
-//    QString escript = QString::fromWCharArray(f->elfScript);
+//    QString escript = QString::fromUtf16((ushort *)f->elfScript);
 //    qDebug("script=%s", escript.latin1());
 
-    NEWTEXTMETRIC *tm = (NEWTEXTMETRIC *)textmetric;
-    fixed = !(tm->tmPitchAndFamily & TMPF_FIXED_PITCH);
-    ttf = (tm->tmPitchAndFamily & TMPF_TRUETYPE);
-    scalable = tm->tmPitchAndFamily & (TMPF_VECTOR|TMPF_TRUETYPE);
-    size = scalable ? SMOOTH_SCALABLE : tm->tmHeight;
-    italic = tm->tmItalic;
-    weight = tm->tmWeight;
-
+    QT_WA({
+        NEWTEXTMETRIC *tm = (NEWTEXTMETRIC *)textmetric;
+        fixed = !(tm->tmPitchAndFamily & TMPF_FIXED_PITCH);
+        ttf = (tm->tmPitchAndFamily & TMPF_TRUETYPE);
+        scalable = tm->tmPitchAndFamily & (TMPF_VECTOR|TMPF_TRUETYPE);
+        size = scalable ? SMOOTH_SCALABLE : tm->tmHeight;
+        italic = tm->tmItalic;
+        weight = tm->tmWeight;
+    } , {
+        NEWTEXTMETRICA *tm = (NEWTEXTMETRICA *)textmetric;
+        fixed = !(tm->tmPitchAndFamily & TMPF_FIXED_PITCH);
+        ttf = (tm->tmPitchAndFamily & TMPF_TRUETYPE);
+        scalable = tm->tmPitchAndFamily & (TMPF_VECTOR|TMPF_TRUETYPE);
+        size = scalable ? SMOOTH_SCALABLE : tm->tmHeight;
+        italic = tm->tmItalic;
+        weight = tm->tmWeight;
+    });
     // the "@family" fonts are just the same as "family". Ignore them.
     if (familyName[0] != QLatin1Char('@') && !familyName.startsWith(QLatin1String("WST_"))) {
         QtFontStyle::Key styleKey;
@@ -385,10 +429,18 @@
 int CALLBACK
 storeFont(ENUMLOGFONTEX* f, NEWTEXTMETRICEX *textmetric, int type, LPARAM /*p*/)
 {
-    QString familyName = QString::fromWCharArray(f->elfLogFont.lfFaceName);
-    QString script = QString::fromWCharArray(f->elfScript);
-
-    FONTSIGNATURE signature = textmetric->ntmFontSig;
+    QString familyName;
+    QT_WA({
+        familyName = QString::fromUtf16((ushort*)f->elfLogFont.lfFaceName);
+    },{
+        ENUMLOGFONTEXA *fa = (ENUMLOGFONTEXA *)f;
+        familyName = QString::fromLocal8Bit(fa->elfLogFont.lfFaceName);
+    });
+    QString script = QT_WA_INLINE(QString::fromUtf16((const ushort *)f->elfScript),
+                                  QString::fromLocal8Bit((const char *)((ENUMLOGFONTEXA *)f)->elfScript));
+
+    FONTSIGNATURE signature;
+    getFontSignature(familyName, textmetric, &signature);
 
     // NEWTEXTMETRICEX is a NEWTEXTMETRIC, which according to the documentation is
     // identical to a TEXTMETRIC except for the last four members, which we don't use
@@ -416,17 +468,33 @@
 
     HDC dummy = GetDC(0);
 
-    LOGFONT lf;
-    lf.lfCharSet = DEFAULT_CHARSET;
-    if (fam.isNull()) {
-        lf.lfFaceName[0] = 0;
-    } else {
-        memcpy(lf.lfFaceName, fam.utf16(), sizeof(wchar_t) * qMin(fam.length() + 1, 32));  // 32 = Windows hard-coded
-    }
-    lf.lfPitchAndFamily = 0;
-
-    EnumFontFamiliesEx(dummy, &lf,
-        (FONTENUMPROC)storeFont, (LPARAM)privateDb(), 0);
+    QT_WA({
+        LOGFONT lf;
+        lf.lfCharSet = DEFAULT_CHARSET;
+        if (fam.isNull()) {
+            lf.lfFaceName[0] = 0;
+        } else {
+            memcpy(lf.lfFaceName, fam.utf16(), sizeof(TCHAR)*qMin(fam.length()+1,32));  // 32 = Windows hard-coded
+        }
+        lf.lfPitchAndFamily = 0;
+
+        EnumFontFamiliesEx(dummy, &lf,
+            (FONTENUMPROC)storeFont, (LPARAM)privateDb(), 0);
+    } , {
+        LOGFONTA lf;
+        lf.lfCharSet = DEFAULT_CHARSET;
+        if (fam.isNull()) {
+            lf.lfFaceName[0] = 0;
+        } else {
+            QByteArray lname = fam.toLocal8Bit();
+            memcpy(lf.lfFaceName,lname.data(),
+                qMin(lname.length()+1,32));  // 32 = Windows hard-coded
+        }
+        lf.lfPitchAndFamily = 0;
+
+        EnumFontFamiliesExA(dummy, &lf,
+            (FONTENUMPROCA)storeFont, (LPARAM)privateDb(), 0);
+    });
 
     ReleaseDC(0, dummy);
 
@@ -437,11 +505,21 @@
         for (int j = 0; j < fnt.families.count(); ++j) {
             const QString familyName = fnt.families.at(j);
             HDC hdc = GetDC(0);
-            LOGFONT lf;
-            memset(&lf, 0, sizeof(LOGFONT));
-            memcpy(lf.lfFaceName, familyName.utf16(), sizeof(wchar_t) * qMin(LF_FACESIZE, familyName.size()));
-            lf.lfCharSet = DEFAULT_CHARSET;
-            HFONT hfont = CreateFontIndirect(&lf);
+            HFONT hfont;
+            QT_WA({
+                LOGFONTW lf;
+                memset(&lf, 0, sizeof(LOGFONTW));
+                memcpy(lf.lfFaceName, familyName.utf16(), qMin(LF_FACESIZE, familyName.size()));
+                lf.lfCharSet = DEFAULT_CHARSET;
+                hfont = CreateFontIndirectW(&lf);
+            } , {
+                LOGFONTA lf;
+                memset(&lf, 0, sizeof(LOGFONTA));
+                QByteArray lfam = familyName.toLocal8Bit();
+                memcpy(lf.lfFaceName, lfam.data(), qMin(LF_FACESIZE, lfam.length()));
+                lf.lfCharSet = DEFAULT_CHARSET;
+                hfont = CreateFontIndirectA(&lf);
+            });
             HGDIOBJ oldobj = SelectObject(hdc, hfont);
 
             TEXTMETRIC textMetrics;
@@ -529,10 +607,17 @@
 
     HDC dc = ((request.styleStrategy & QFont::PreferDevice) && fontHdc) ? fontHdc : shared_dc();
     SelectObject(dc, fe->hfont);
-    wchar_t n[64];
-    GetTextFace(dc, 64, n);
-    fe->fontDef.family = QString::fromWCharArray(n);
-    fe->fontDef.fixedPitch = !(fe->tm.tmPitchAndFamily & TMPF_FIXED_PITCH);
+    QT_WA({
+        TCHAR n[64];
+        GetTextFaceW(dc, 64, n);
+        fe->fontDef.family = QString::fromUtf16((ushort*)n);
+        fe->fontDef.fixedPitch = !(fe->tm.w.tmPitchAndFamily & TMPF_FIXED_PITCH);
+    } , {
+        char an[64];
+        GetTextFaceA(dc, 64, an);
+        fe->fontDef.family = QString::fromLocal8Bit(an);
+        fe->fontDef.fixedPitch = !(fe->tm.a.tmPitchAndFamily & TMPF_FIXED_PITCH);
+    });
     if (fe->fontDef.pointSize < 0) {
         fe->fontDef.pointSize = fe->fontDef.pixelSize * 72. / dpi;
     } else if (fe->fontDef.pixelSize == -1) {
@@ -774,7 +859,11 @@
         } else if (request.styleStrategy & QFont::PreferDevice) {
             strat = OUT_DEVICE_PRECIS;
         } else if (request.styleStrategy & QFont::PreferOutline) {
-            strat = OUT_OUTLINE_PRECIS;
+            QT_WA({
+                strat = OUT_OUTLINE_PRECIS;
+            } , {
+                strat = OUT_TT_PRECIS;
+            });
         } else if (request.styleStrategy & QFont::ForceOutline) {
             strat = OUT_TT_ONLY_PRECIS;
 #endif
@@ -793,7 +882,7 @@
 
         if (request.styleStrategy & QFont::PreferAntialias) {
             if (QSysInfo::WindowsVersion >= QSysInfo::WV_XP) {
-                qual = CLEARTYPE_QUALITY;
+                qual = 5; // == CLEARTYPE_QUALITY;
                 preferClearTypeAA = true;
             } else {
                 qual = ANTIALIASED_QUALITY;
@@ -819,9 +908,16 @@
         if (fam == QLatin1String("Courier") && !(request.styleStrategy & QFont::PreferBitmap))
             fam = QLatin1String("Courier New");
 
-        memcpy(lf.lfFaceName, fam.utf16(), sizeof(wchar_t) * qMin(fam.length() + 1, 32));  // 32 = Windows hard-coded
-
-        hfont = CreateFontIndirect(&lf);
+        QT_WA({
+            memcpy(lf.lfFaceName, fam.utf16(), sizeof(TCHAR)*qMin(fam.length()+1,32));  // 32 = Windows hard-coded
+            hfont = CreateFontIndirect(&lf);
+        } , {
+            // LOGFONTA and LOGFONTW are binary compatible
+            QByteArray lname = fam.toLocal8Bit();
+            memcpy(lf.lfFaceName,lname.data(),
+                qMin(lname.length()+1,32));  // 32 = Windows hard-coded
+            hfont = CreateFontIndirectA((LOGFONTA*)&lf);
+        });
         if (!hfont)
             qErrnoWarning("QFontEngine::loadEngine: CreateFontIndirect failed");
 
@@ -830,11 +926,17 @@
         int avWidth = 0;
         BOOL res;
         HGDIOBJ oldObj = SelectObject(hdc, hfont);
-
-        TEXTMETRIC tm;
-        res = GetTextMetrics(hdc, &tm);
-        avWidth = tm.tmAveCharWidth;
-        ttf = tm.tmPitchAndFamily & TMPF_TRUETYPE;
+        QT_WA({
+            TEXTMETRICW tm;
+            res = GetTextMetricsW(hdc, &tm);
+            avWidth = tm.tmAveCharWidth;
+            ttf = tm.tmPitchAndFamily & TMPF_TRUETYPE;
+        } , {
+            TEXTMETRICA tm;
+            res = GetTextMetricsA(hdc, &tm);
+            avWidth = tm.tmAveCharWidth;
+            ttf = tm.tmPitchAndFamily & TMPF_TRUETYPE;
+        });
         SelectObject(hdc, oldObj);
 
         if (!ttf || !useDirectWrite) {
@@ -845,7 +947,7 @@
                 if (!res)
                     qErrnoWarning("QFontEngine::loadEngine: GetTextMetrics failed");
                 lf.lfWidth = avWidth * request.stretch/100;
-                hfont = CreateFontIndirect(&lf);
+                hfont = QT_WA_INLINE(CreateFontIndirectW(&lf), CreateFontIndirectA((LOGFONTA *)&lf));
                 if (!hfont)
                     qErrnoWarning("QFontEngine::loadEngine: CreateFontIndirect with stretch failed");
             }
@@ -1035,6 +1137,12 @@
     // list of families to try
     QStringList family_list = familyList(req);
 
+    if(QSysInfo::WindowsVersion & QSysInfo::WV_DOS_based && req.family.toLower() == QLatin1String("ms sans serif")) {
+        // small hack for Dos based machines to get the right font for non
+        // latin text when using the default font.
+        family_list << QLatin1String("Arial");
+    }
+
     const char *stylehint = styleHint(d->request);
     if (stylehint)
         family_list << QLatin1String(stylehint);
@@ -1228,21 +1336,21 @@
         HANDLE handle = 0;
 
         {
-#ifdef QT_NO_TEMPORARYFILE
-           wchar_t lpBuffer[MAX_PATH];
+#ifdef  QT_NO_TEMPORARYFILE
+           TCHAR lpBuffer[MAX_PATH];
            GetTempPath(MAX_PATH, lpBuffer);
-           QString s = QString::fromWCharArray(lpBuffer);
+           QString s = QString::fromUtf16((const ushort *) lpBuffer);
            QFile tempfile(s + QLatin1String("/font") + QString::number(GetTickCount()) + QLatin1String(".ttf"));
            if (!tempfile.open(QIODevice::ReadWrite))
 #else
             QTemporaryFile tempfile(QLatin1String("XXXXXXXX.ttf"));
             if (!tempfile.open())
-#endif // QT_NO_TEMPORARYFILE
+#endif
                 return;
             if (tempfile.write(fnt->data) == -1)
                 return;
 
-#ifndef QT_NO_TEMPORARYFILE
+#ifndef  QT_NO_TEMPORARYFILE
             tempfile.setAutoRemove(false);
 #endif
             fnt->fileName = QFileInfo(tempfile.fileName()).absoluteFilePath();
@@ -1254,11 +1362,13 @@
         }
 #else
         DWORD dummy = 0;
-        HANDLE handle = ptrAddFontMemResourceEx((void *)fnt->data.constData(), fnt->data.size(), 0,
-                                                &dummy);
+        HANDLE handle = ptrAddFontMemResourceEx((void *)fnt->data.constData(),
+                                    fnt->data.size(),
+                                    0,
+                                    &dummy);
         if (handle == 0)
             return;
-#endif // Q_OS_WINCE
+#endif
 
         fnt->handle = handle;
         fnt->data = QByteArray();
@@ -1279,10 +1389,12 @@
 #else
         PtrAddFontResourceExW ptrAddFontResourceExW = (PtrAddFontResourceExW)QSystemLibrary::resolve(QLatin1String("gdi32"),
                                                                                                "AddFontResourceExW");
-        if (!ptrAddFontResourceExW
-            || ptrAddFontResourceExW((wchar_t*)fnt->fileName.utf16(), FR_PRIVATE, 0) == 0)
+        if (!ptrAddFontResourceExW)
             return;
-#endif // Q_OS_WINCE
+
+        if (ptrAddFontResourceExW((LPCWSTR)fnt->fileName.utf16(), FR_PRIVATE, 0) == 0)
+            return;
+#endif
 
         fnt->memoryFont = false;
     }
@@ -1308,10 +1420,12 @@
 #else
         PtrRemoveFontMemResourceEx ptrRemoveFontMemResourceEx = (PtrRemoveFontMemResourceEx)QSystemLibrary::resolve(QLatin1String("gdi32"),
                                                                                                               "RemoveFontMemResourceEx");
-        if (!ptrRemoveFontMemResourceEx
-            || !ptrRemoveFontMemResourceEx(font.handle))
+        if (!ptrRemoveFontMemResourceEx)
             return false;
-#endif // Q_OS_WINCE
+
+        if (!ptrRemoveFontMemResourceEx(font.handle))
+            return false;
+#endif
     } else {
 #ifdef Q_OS_WINCE
         if (!RemoveFontResource((LPCWSTR)font.fileName.utf16()))
@@ -1319,10 +1433,12 @@
 #else
         PtrRemoveFontResourceExW ptrRemoveFontResourceExW = (PtrRemoveFontResourceExW)QSystemLibrary::resolve(QLatin1String("gdi32"),
                                                                                                         "RemoveFontResourceExW");
-        if (!ptrRemoveFontResourceExW
-            || !ptrRemoveFontResourceExW((LPCWSTR)font.fileName.utf16(), FR_PRIVATE, 0))
+        if (!ptrRemoveFontResourceExW)
             return false;
-#endif // Q_OS_WINCE
+
+        if (!ptrRemoveFontResourceExW((LPCWSTR)font.fileName.utf16(), FR_PRIVATE, 0))
+            return false;
+#endif
     }
 
     db->invalidate();
--- a/src/gui/text/qfontengine_win.cpp
+++ b/src/gui/text/qfontengine_win.cpp
@@ -53,6 +53,7 @@
 #include <private/qsystemlibrary_p.h>
 #include <qpaintdevice.h>
 #include <qpainter.h>
+#include <qlibrary.h>
 #include <limits.h>
 
 #include <qendian.h>
@@ -88,6 +89,8 @@
     ((quint32)(ch1)) \
    )
 
+typedef BOOL (WINAPI *PtrGetCharWidthI)(HDC, UINT, UINT, LPWORD, LPINT);
+
 // common DC for all fonts
 
 QT_BEGIN_NAMESPACE
@@ -129,6 +132,7 @@
 }
 #endif
 
+static HFONT stock_sysfont = 0;
 #ifndef Q_WS_WINCE
 typedef BOOL (WINAPI *PtrGetCharWidthI)(HDC, UINT, UINT, LPWORD, LPINT);
 static PtrGetCharWidthI ptrGetCharWidthI = 0;
@@ -145,6 +149,28 @@
     resolvedGetCharWidthI = true;
 }
 #endif // !defined(Q_WS_WINCE)
+
+// Copy a LOGFONTW struct into a LOGFONTA by converting the face name to an 8 bit value.
+// This is needed when calling CreateFontIndirect on non-unicode windowses.
+inline static void wa_copy_logfont(LOGFONTW *lfw, LOGFONTA *lfa)
+{
+    lfa->lfHeight = lfw->lfHeight;
+    lfa->lfWidth = lfw->lfWidth;
+    lfa->lfEscapement = lfw->lfEscapement;
+    lfa->lfOrientation = lfw->lfOrientation;
+    lfa->lfWeight = lfw->lfWeight;
+    lfa->lfItalic = lfw->lfItalic;
+    lfa->lfUnderline = lfw->lfUnderline;
+    lfa->lfStrikeOut = lfw->lfStrikeOut;
+    lfa->lfCharSet = lfw->lfCharSet;
+    lfa->lfOutPrecision = lfw->lfOutPrecision;
+    lfa->lfClipPrecision = lfw->lfClipPrecision;
+    lfa->lfQuality = lfw->lfQuality;
+    lfa->lfPitchAndFamily = lfw->lfPitchAndFamily;
+
+    QString fam = QString::fromUtf16((const ushort*)lfw->lfFaceName);
+    qstrncpy(lfa->lfFaceName, fam.toLocal8Bit().constData(), LF_FACESIZE);
+}
 
 // defined in qtextengine_win.cpp
 typedef void *SCRIPT_CACHE;
@@ -171,6 +197,14 @@
     return val;
 }
 
+static inline HFONT systemFont()
+{
+    if (stock_sysfont == 0)
+        stock_sysfont = (HFONT)GetStockObject(SYSTEM_FONT);
+    return stock_sysfont;
+}
+
+
 // general font engine
 
 QFixed QFontEngineWin::lineThickness() const
@@ -181,14 +215,25 @@
     return QFontEngine::lineThickness();
 }
 
-static OUTLINETEXTMETRIC *getOutlineTextMetric(HDC hdc)
+#if defined(Q_WS_WINCE)
+static OUTLINETEXTMETRICW *getOutlineTextMetric(HDC hdc)
 {
     int size;
-    size = GetOutlineTextMetrics(hdc, 0, 0);
-    OUTLINETEXTMETRIC *otm = (OUTLINETEXTMETRIC *)malloc(size);
-    GetOutlineTextMetrics(hdc, size, otm);
+    size = GetOutlineTextMetricsW(hdc, 0, 0);
+    OUTLINETEXTMETRICW *otm = (OUTLINETEXTMETRICW *)malloc(size);
+    GetOutlineTextMetricsW(hdc, size, otm);
     return otm;
 }
+#else
+static OUTLINETEXTMETRICA *getOutlineTextMetric(HDC hdc)
+{
+    int size;
+    size = GetOutlineTextMetricsA(hdc, 0, 0);
+    OUTLINETEXTMETRICA *otm = (OUTLINETEXTMETRICA *)malloc(size);
+    GetOutlineTextMetricsA(hdc, size, otm);
+    return otm;
+}
+#endif
 
 bool QFontEngineWin::hasCFFTable() const
 {
@@ -206,7 +251,7 @@
 
 void QFontEngineWin::getCMap()
 {
-    ttf = (bool)(tm.tmPitchAndFamily & TMPF_TRUETYPE) || hasCMapTable();
+    ttf = (bool)(QT_WA_INLINE(tm.w.tmPitchAndFamily, tm.a.tmPitchAndFamily) & TMPF_TRUETYPE) || hasCMapTable();
 
     cffTable = hasCFFTable();
 
@@ -227,7 +272,11 @@
     designToDevice = 1;
     _faceId.index = 0;
     if(cmap) {
-        OUTLINETEXTMETRIC *otm = getOutlineTextMetric(hdc);
+#if defined(Q_WS_WINCE)
+        OUTLINETEXTMETRICW *otm = getOutlineTextMetric(hdc);
+#else
+        OUTLINETEXTMETRICA *otm = getOutlineTextMetric(hdc);
+#endif
         designToDevice = QFixed((int)otm->otmEMSquare)/int(otm->otmTextMetrics.tmHeight);
         unitsPerEm = otm->otmEMSquare;
         x_height = (int)otm->otmsXHeight;
@@ -237,7 +286,7 @@
         fsType = otm->otmfsType;
         free(otm);
     } else {
-        unitsPerEm = tm.tmHeight;
+        unitsPerEm = tm.w.tmHeight;
     }
 }
 
@@ -274,14 +323,19 @@
             }
         } else {
 #endif
-            wchar_t first = tm.tmFirstChar;
-            wchar_t last = tm.tmLastChar;
-
+            ushort first, last;
+            QT_WA({
+                first = tm.w.tmFirstChar;
+                last = tm.w.tmLastChar;
+            }, {
+                first = tm.a.tmFirstChar;
+                last = tm.a.tmLastChar;
+            });
             for (; i < numChars; ++i, ++glyph_pos) {
                 uint ucs = QChar::mirroredChar(getChar(str, i, numChars));
                 if (
 #ifdef Q_WS_WINCE
-                    tm.tmFirstChar > 60000 || // see line 375
+                    tm.w.tmFirstChar > 60000 || // see line 375
 #endif
                         ucs >= first && ucs <= last)
                     glyphs->glyphs[glyph_pos] = ucs;
@@ -307,14 +361,19 @@
             }
         } else {
 #endif
-            wchar_t first = tm.tmFirstChar;
-            wchar_t last = tm.tmLastChar;
-
+            ushort first, last;
+            QT_WA({
+                first = tm.w.tmFirstChar;
+                last = tm.w.tmLastChar;
+            }, {
+                first = tm.a.tmFirstChar;
+                last = tm.a.tmLastChar;
+            });
             for (; i < numChars; ++i, ++glyph_pos) {
                 uint uc = getChar(str, i, numChars);
                 if (
 #ifdef Q_WS_WINCE
-                    tm.tmFirstChar > 60000 || // see comment in QFontEngineWin
+                    tm.w.tmFirstChar > 60000 || // see comment in QFontEngineWin
 #endif
                         uc >= first && uc <= last)
                     glyphs->glyphs[glyph_pos] = uc;
@@ -348,16 +407,25 @@
     lineWidth = -1;
     x_height = -1;
 
-    BOOL res = GetTextMetrics(hdc, &tm);
-    fontDef.fixedPitch = !(tm.tmPitchAndFamily & TMPF_FIXED_PITCH);
+    BOOL res = QT_WA_INLINE(GetTextMetricsW(hdc, &tm.w), GetTextMetricsA(hdc, &tm.a));
+    fontDef.fixedPitch = !(QT_WA_INLINE(tm.w.tmPitchAndFamily, tm.a.tmPitchAndFamily) & TMPF_FIXED_PITCH);
     if (!res) {
         qErrnoWarning("QFontEngineWin: GetTextMetrics failed");
         ZeroMemory(&tm, sizeof(TEXTMETRIC));
     }
 
-    cache_cost = tm.tmHeight * tm.tmAveCharWidth * 2000;
+    cache_cost = tm.w.tmHeight * tm.w.tmAveCharWidth * 2000;
     getCMap();
 
+    useTextOutA = false;
+#ifndef Q_OS_WINCE
+    // TextOutW doesn't work for symbol fonts on Windows 95!
+    // since we're using glyph indices we don't care for ttfs about this!
+    if (QSysInfo::WindowsVersion == QSysInfo::WV_95 && !ttf &&
+         (_name == QLatin1String("Marlett") || _name == QLatin1String("Symbol") ||
+           _name == QLatin1String("Webdings") || _name == QLatin1String("Wingdings")))
+            useTextOutA = true;
+#endif
     widthCache = 0;
     widthCacheSize = 0;
     designAdvances = 0;
@@ -378,7 +446,7 @@
         free(widthCache);
 
     // make sure we aren't by accident still selected
-    SelectObject(shared_dc(), (HFONT)GetStockObject(SYSTEM_FONT));
+    SelectObject(shared_dc(), systemFont());
 
     if (!stockFont) {
         if (!DeleteObject(hfont))
@@ -386,13 +454,40 @@
     }
 }
 
-HGDIOBJ QFontEngineWin::selectDesignFont() const
+HGDIOBJ QFontEngineWin::selectDesignFont(QFixed *overhang) const
 {
     LOGFONT f = logfont;
     f.lfHeight = unitsPerEm;
     f.lfWidth = 0;
-    HFONT designFont = CreateFontIndirect(&f);
-    return SelectObject(shared_dc(), designFont);
+    HFONT designFont;
+    QT_WA({
+        designFont = CreateFontIndirectW(&f);
+    }, {
+        LOGFONTA fa;
+        wa_copy_logfont(&f, &fa);
+        designFont = CreateFontIndirectA(&fa);
+    });
+    HGDIOBJ oldFont = SelectObject(shared_dc(), designFont);
+
+    if (QSysInfo::WindowsVersion & QSysInfo::WV_DOS_based) {
+        BOOL res;
+        QT_WA({
+            TEXTMETRICW tm;
+            res = GetTextMetricsW(shared_dc(), &tm);
+            if (!res)
+                qErrnoWarning("QFontEngineWin: GetTextMetrics failed");
+            *overhang = QFixed((int)tm.tmOverhang) / designToDevice;
+        } , {
+            TEXTMETRICA tm;
+            res = GetTextMetricsA(shared_dc(), &tm);
+            if (!res)
+                qErrnoWarning("QFontEngineWin: GetTextMetrics failed");
+            *overhang = QFixed((int)tm.tmOverhang) / designToDevice;
+        });
+    } else {
+        *overhang = 0;
+    }
+    return oldFont;
 }
 
 bool QFontEngineWin::stringToCMap(const QChar *str, int len, QGlyphLayout *glyphs, int *nglyphs, QTextEngine::ShaperFlags flags) const
@@ -416,8 +511,18 @@
 #if defined(Q_WS_WINCE)
     GetCharWidth32(hdc, glyph, glyph, &width);
 #else
-    if (ptrGetCharWidthI)
+    if (ptrGetCharWidthI) {
         ptrGetCharWidthI(hdc, glyph, 1, 0, &width);
+    } else {
+        MAT2 mat;
+        memset(&mat, 0, sizeof(mat));
+        mat.eM11.value = mat.eM22.value = 1;
+        GLYPHMETRICS metrics;
+        DWORD result = QT_WA_INLINE(GetGlyphOutlineW(hdc, glyph, GGO_METRICS | GGO_GLYPH_INDEX, &metrics, 0, 0, &mat),
+                GetGlyphOutlineA(hdc, glyph, GGO_METRICS | GGO_GLYPH_INDEX, &metrics, 0, 0, &mat));
+        if (result != GDI_ERROR)
+            width = metrics.gmCellIncX;
+    }
 #endif
 }
 
@@ -426,6 +531,8 @@
     HGDIOBJ oldFont = 0;
     HDC hdc = shared_dc();
     if (ttf && (flags & QTextEngine::DesignMetrics)) {
+        QFixed overhang = 0;
+
         for(int i = 0; i < glyphs->numGlyphs; i++) {
             unsigned int glyph = glyphs->glyphs[i];
             if(int(glyph) >= designAdvancesSize) {
@@ -436,9 +543,9 @@
                     designAdvances[i] = -1000000;
                 designAdvancesSize = newSize;
             }
-            if (designAdvances[glyph] < -999999) {
-                if (!oldFont)
-                    oldFont = selectDesignFont();
+            if(designAdvances[glyph] < -999999) {
+                if(!oldFont)
+                    oldFont = selectDesignFont(&overhang);
 
                 int width = 0;
                 calculateTTFGlyphWidth(hdc, glyph, width);
@@ -450,6 +557,8 @@
         if(oldFont)
             DeleteObject(SelectObject(hdc, oldFont));
     } else {
+        int overhang = (QSysInfo::WindowsVersion & QSysInfo::WV_DOS_based) ? tm.a.tmOverhang : 0;
+
         for(int i = 0; i < glyphs->numGlyphs; i++) {
             unsigned int glyph = glyphs->glyphs[i];
 
@@ -478,7 +587,7 @@
                         ++chrLen;
                     }
                     SIZE size = {0, 0};
-                    GetTextExtentPoint32(hdc, (wchar_t *)ch, chrLen, &size);
+                    GetTextExtentPoint32W(hdc, (wchar_t *)ch, chrLen, &size);
                     width = size.cx;
                 } else {
                     calculateTTFGlyphWidth(hdc, glyph, width);
@@ -504,7 +613,7 @@
     for (int i = 0; i < glyphs.numGlyphs; ++i)
         w += glyphs.effectiveAdvance(i);
 
-    return glyph_metrics_t(0, -tm.tmAscent, w - lastRightBearing(glyphs), tm.tmHeight, w, 0);
+    return glyph_metrics_t(0, -tm.w.tmAscent, w - lastRightBearing(glyphs), tm.w.tmHeight, w, 0);
 }
 
 #ifndef Q_WS_WINCE
@@ -541,7 +650,7 @@
     uint format = GGO_METRICS;
     if (ttf)
         format |= GGO_GLYPH_INDEX;
-    res = GetGlyphOutline(hdc, glyph, format, &gm, 0, 0, &mat);
+    res = QT_WA_INLINE(GetGlyphOutlineW(hdc, glyph, format, &gm, 0, 0, &mat), GetGlyphOutlineA(hdc, glyph, format, &gm, 0, 0, &mat));
 
     if (t.type() > QTransform::TxTranslate) {
         XFORM xform;
@@ -573,11 +682,17 @@
     if (!ttf && !success) {
         // Bitmap fonts
         wchar_t ch = glyph;
+        if (QSysInfo::WindowsVersion & QSysInfo::WV_DOS_based) {
+            SIZE size = { 0, 0 };
+            GetTextExtentPoint32W(hdc, &ch, 1, &size);
+            return glyph_metrics_t(0, -tm.a.tmAscent, size.cx, tm.a.tmHeight,
+                    size.cx - tm.a.tmOverhang, 0).transformed(t);
+        }
         ABCFLOAT abc;
-        GetCharABCWidthsFloat(hdc, ch, ch, &abc);
+        GetCharABCWidthsFloatW(hdc, ch, ch, &abc);
         int width = qRound(abc.abcfB);
 
-        return glyph_metrics_t(QFixed::fromReal(abc.abcfA), -tm.tmAscent, width, tm.tmHeight, width, 0).transformed(t);
+        return glyph_metrics_t(QFixed::fromReal(abc.abcfA), -tm.w.tmAscent, width, tm.w.tmHeight, width, 0).transformed(t);
     }
 
     return glyphMetrics;
@@ -602,30 +717,30 @@
     else
 #endif
     {   // fallback
-        width = tm.tmMaxCharWidth;
+        width = tm.w.tmMaxCharWidth;
         advance = width;
     }
 
     SelectObject(hdc, oldFont);
-    return glyph_metrics_t(0, -tm.tmAscent, width, tm.tmHeight, advance, 0).transformed(t);
+    return glyph_metrics_t(0, -tm.w.tmAscent, width, tm.w.tmHeight, advance, 0).transformed(t);
 #endif
 }
 
 QFixed QFontEngineWin::ascent() const
 {
-    return tm.tmAscent;
+    return tm.w.tmAscent;
 }
 
 QFixed QFontEngineWin::descent() const
 {
     // ### we substract 1 to even out the historical +1 in QFontMetrics's
     // ### height=asc+desc+1 equation. Fix in Qt5.
-    return tm.tmDescent - 1;
+    return tm.w.tmDescent - 1;
 }
 
 QFixed QFontEngineWin::leading() const
 {
-    return tm.tmExternalLeading;
+    return tm.w.tmExternalLeading;
 }
 
 
@@ -638,12 +753,12 @@
 
 QFixed QFontEngineWin::averageCharWidth() const
 {
-    return tm.tmAveCharWidth;
+    return tm.w.tmAveCharWidth;
 }
 
 qreal QFontEngineWin::maxCharWidth() const
 {
-    return tm.tmMaxCharWidth;
+    return tm.w.tmMaxCharWidth;
 }
 
 enum { max_font_count = 256 };
@@ -716,10 +831,10 @@
         SelectObject(hdc, hfont);
         if (ttf) {
             ABC *abc = 0;
-            int n = tm.tmLastChar - tm.tmFirstChar;
+            int n = QT_WA_INLINE(tm.w.tmLastChar - tm.w.tmFirstChar, tm.a.tmLastChar - tm.a.tmFirstChar);
             if (n <= max_font_count) {
                 abc = new ABC[n+1];
-                GetCharABCWidths(hdc, tm.tmFirstChar, tm.tmLastChar, abc);
+                GetCharABCWidths(hdc, tm.w.tmFirstChar, tm.w.tmLastChar, abc);
             } else {
                 abc = new ABC[char_table_entries+1];
                 for(int i = 0; i < char_table_entries; i++)
@@ -735,6 +850,9 @@
                 }
             }
             delete [] abc;
+        } else {
+            ml = 0;
+            mr = -tm.a.tmOverhang;
         }
         lbearing = ml;
         rbearing = mr;
@@ -749,14 +867,28 @@
         SelectObject(hdc, hfont);
         if (ttf) {
             ABC *abc = 0;
-            int n = tm.tmLastChar - tm.tmFirstChar;
+            int n = QT_WA_INLINE(tm.w.tmLastChar - tm.w.tmFirstChar, tm.a.tmLastChar - tm.a.tmFirstChar);
             if (n <= max_font_count) {
                 abc = new ABC[n+1];
-                GetCharABCWidths(hdc, tm.tmFirstChar, tm.tmLastChar, abc);
+                QT_WA({
+                    GetCharABCWidths(hdc, tm.w.tmFirstChar, tm.w.tmLastChar, abc);
+                }, {
+                    GetCharABCWidthsA(hdc,tm.a.tmFirstChar,tm.a.tmLastChar,abc);
+                });
             } else {
                 abc = new ABC[char_table_entries+1];
-                for(int i = 0; i < char_table_entries; i++)
-                    GetCharABCWidths(hdc, char_table[i], char_table[i], abc + i);
+                QT_WA({
+                    for(int i = 0; i < char_table_entries; i++)
+                        GetCharABCWidths(hdc, char_table[i], char_table[i], abc+i);
+                }, {
+                    for(int i = 0; i < char_table_entries; i++) {
+                        QByteArray w = QString(QChar(char_table[i])).toLocal8Bit();
+                        if (w.length() == 1) {
+                            uint ch8 = (uchar)w[0];
+                            GetCharABCWidthsA(hdc, ch8, ch8, abc+i);
+                        }
+                    }
+                });
                 n = char_table_entries;
             }
             ml = abc[0].abcA;
@@ -769,28 +901,33 @@
             }
             delete [] abc;
         } else {
-            ABCFLOAT *abc = 0;
-            int n = tm.tmLastChar - tm.tmFirstChar+1;
-            if (n <= max_font_count) {
-                abc = new ABCFLOAT[n];
-                GetCharABCWidthsFloat(hdc, tm.tmFirstChar, tm.tmLastChar, abc);
-            } else {
-                abc = new ABCFLOAT[char_table_entries];
-                for(int i = 0; i < char_table_entries; i++)
-                    GetCharABCWidthsFloat(hdc, char_table[i], char_table[i], abc+i);
-                n = char_table_entries;
-            }
-            float fml = abc[0].abcfA;
-            float fmr = abc[0].abcfC;
-            for (int i=1; i<n; i++) {
-                if (abc[i].abcfA + abc[i].abcfB + abc[i].abcfC != 0) {
-                    fml = qMin(fml,abc[i].abcfA);
-                    fmr = qMin(fmr,abc[i].abcfC);
+            QT_WA({
+                ABCFLOAT *abc = 0;
+                int n = tm.w.tmLastChar - tm.w.tmFirstChar+1;
+                if (n <= max_font_count) {
+                    abc = new ABCFLOAT[n];
+                    GetCharABCWidthsFloat(hdc, tm.w.tmFirstChar, tm.w.tmLastChar, abc);
+                } else {
+                    abc = new ABCFLOAT[char_table_entries];
+                    for(int i = 0; i < char_table_entries; i++)
+                        GetCharABCWidthsFloat(hdc, char_table[i], char_table[i], abc+i);
+                    n = char_table_entries;
                 }
-            }
-            ml = int(fml - 0.9999);
-            mr = int(fmr - 0.9999);
-            delete [] abc;
+                float fml = abc[0].abcfA;
+                float fmr = abc[0].abcfC;
+                for (int i=1; i<n; i++) {
+                    if (abc[i].abcfA + abc[i].abcfB + abc[i].abcfC != 0) {
+                        fml = qMin(fml,abc[i].abcfA);
+                        fmr = qMin(fmr,abc[i].abcfC);
+                    }
+                }
+                ml = int(fml-0.9999);
+                mr = int(fmr-0.9999);
+                delete [] abc;
+            } , {
+                ml = 0;
+                mr = -tm.a.tmOverhang;
+            });
         }
         lbearing = ml;
         rbearing = mr;
@@ -827,10 +964,17 @@
                 return false;
         }
     } else {
-        while(len--) {
-            if (tm.tmFirstChar > string->unicode() || tm.tmLastChar < string->unicode())
-                return false;
-        }
+        QT_WA({
+            while(len--) {
+                if (tm.w.tmFirstChar > string->unicode() || tm.w.tmLastChar < string->unicode())
+                    return false;
+            }
+        }, {
+            while(len--) {
+                if (tm.a.tmFirstChar > string->unicode() || tm.a.tmLastChar < string->unicode())
+                    return false;
+            }
+        });
     }
     return true;
 }
@@ -876,7 +1020,7 @@
         uint format = GGO_METRICS;
         if (ttf)
             format |= GGO_GLYPH_INDEX;
-        int res = GetGlyphOutline(hdc, glyph, format, &gMetric, 0, 0, &mat);
+        int res = QT_WA_INLINE(GetGlyphOutlineW(hdc, glyph, format, &gMetric, 0, 0, &mat), GetGlyphOutlineA(hdc, glyph, format, &gMetric, 0, 0, &mat));
         if (res == GDI_ERROR) {
             return false;
         }
@@ -894,7 +1038,11 @@
 
     int bufferSize = GDI_ERROR;
 #if !defined(Q_WS_WINCE)
-    bufferSize = GetGlyphOutline(hdc, glyph, glyphFormat, &gMetric, 0, 0, &mat);
+    QT_WA( {
+        bufferSize = GetGlyphOutlineW(hdc, glyph, glyphFormat, &gMetric, 0, 0, &mat);
+    }, {
+        bufferSize = GetGlyphOutlineA(hdc, glyph, glyphFormat, &gMetric, 0, 0, &mat);
+    });
 #endif
     if ((DWORD)bufferSize == GDI_ERROR) {
         return false;
@@ -903,7 +1051,13 @@
     void *dataBuffer = new char[bufferSize];
     DWORD ret = GDI_ERROR;
 #if !defined(Q_WS_WINCE)
-    ret = GetGlyphOutline(hdc, glyph, glyphFormat, &gMetric, bufferSize, dataBuffer, &mat);
+    QT_WA( {
+        ret = GetGlyphOutlineW(hdc, glyph, glyphFormat, &gMetric, bufferSize,
+                            dataBuffer, &mat);
+    }, {
+        ret = GetGlyphOutlineA(hdc, glyph, glyphFormat, &gMetric, bufferSize,
+                                dataBuffer, &mat);
+    } );
 #endif
     if (ret == GDI_ERROR) {
         delete [](char *)dataBuffer;
@@ -992,7 +1146,14 @@
     // font at the correct pixel size.
     lf.lfHeight = -unitsPerEm;
     lf.lfWidth = 0;
-    HFONT hf = CreateFontIndirect(&lf);
+    HFONT hf;
+    QT_WA({
+        hf = CreateFontIndirectW(&lf);
+    }, {
+        LOGFONTA lfa;
+        wa_copy_logfont(&lf, &lfa);
+        hf = CreateFontIndirectA(&lfa);
+    });
     HDC hdc = shared_dc();
     HGDIOBJ oldfont = SelectObject(hdc, hf);
 
@@ -1014,7 +1175,7 @@
                                       QPainterPath *path, QTextItem::RenderFlags flags)
 {
 #if !defined(Q_WS_WINCE)
-    if(tm.tmPitchAndFamily & (TMPF_TRUETYPE | TMPF_VECTOR)) {
+    if(tm.w.tmPitchAndFamily & (TMPF_TRUETYPE | TMPF_VECTOR)) {
         hasOutline = true;
         QFontEngine::addOutlineToPath(x, y, glyphs, path, flags);
         if (hasOutline)  {
@@ -1048,11 +1209,11 @@
             uchar data[4];
             GetFontData(hdc, HEAD, 44, &data, 4);
             USHORT macStyle = getUShort(data);
-            if (tm.tmItalic && !(macStyle & 2))
+            if (tm.w.tmItalic && !(macStyle & 2))
                 synthesized_flags = SynthesizedItalic;
             if (fontDef.stretch != 100 && ttf)
                 synthesized_flags |= SynthesizedStretch;
-            if (tm.tmWeight >= 500 && !(macStyle & 1))
+            if (tm.w.tmWeight >= 500 && !(macStyle & 1))
                 synthesized_flags |= SynthesizedBold;
             //qDebug() << "font is" << _name <<
             //    "it=" << (macStyle & 2) << fontDef.style << "flags=" << synthesized_flags;
@@ -1068,12 +1229,24 @@
 
 QFontEngine::Properties QFontEngineWin::properties() const
 {
+
     LOGFONT lf = logfont;
     lf.lfHeight = unitsPerEm;
-    HFONT hf = CreateFontIndirect(&lf);
+    HFONT hf;
+    QT_WA({
+        hf = CreateFontIndirectW(&lf);
+    }, {
+        LOGFONTA lfa;
+        wa_copy_logfont(&lf, &lfa);
+        hf = CreateFontIndirectA(&lfa);
+    });
     HDC hdc = shared_dc();
     HGDIOBJ oldfont = SelectObject(hdc, hf);
-    OUTLINETEXTMETRIC *otm = getOutlineTextMetric(hdc);
+#if defined(Q_WS_WINCE)
+    OUTLINETEXTMETRICW *otm = getOutlineTextMetric(hdc);
+#else
+    OUTLINETEXTMETRICA *otm = getOutlineTextMetric(hdc);
+#endif
     Properties p;
     p.emSquare = unitsPerEm;
     p.italicAngle = otm->otmItalicAngle;
@@ -1101,7 +1274,14 @@
     if(flags & SynthesizedItalic)
         lf.lfItalic = false;
     lf.lfWidth = 0;
-    HFONT hf = CreateFontIndirect(&lf);
+    HFONT hf;
+    QT_WA({
+        hf = CreateFontIndirectW(&lf);
+    }, {
+        LOGFONTA lfa;
+        wa_copy_logfont(&lf, &lfa);
+        hf = CreateFontIndirectA(&lfa);
+    });
     HDC hdc = shared_dc();
     HGDIOBJ oldfont = SelectObject(hdc, hf);
     QFixedPoint p;
@@ -1171,7 +1351,7 @@
         memset(&mat, 0, sizeof(mat));
         mat.eM11.value = mat.eM22.value = 1;
 
-        if (GetGlyphOutline(hdc, glyph, ggo_options, &tgm, 0, 0, &mat) == GDI_ERROR) {
+        if (QT_WA_INLINE(GetGlyphOutlineW(hdc, glyph, ggo_options, &tgm, 0, 0, &mat), GetGlyphOutlineA(hdc, glyph, ggo_options, &tgm, 0, 0, &mat)) == GDI_ERROR) {
             qWarning("QWinFontEngine: unable to query transformed glyph metrics...");
             return 0;
         }
@@ -1217,11 +1397,11 @@
     if (has_transformation) {
         SetGraphicsMode(hdc, GM_ADVANCED);
         SetWorldTransform(hdc, &xform);
-        ExtTextOut(hdc, 0, 0, options, 0, (LPCWSTR) &glyph, 1, 0);
+        ExtTextOutW(hdc, 0, 0, options, 0, (LPCWSTR) &glyph, 1, 0);
     } else
 #endif
     {
-        ExtTextOut(hdc, -gx + margin, -gy + margin, options, 0, (LPCWSTR) &glyph, 1, 0);
+        ExtTextOutW(hdc, -gx + margin, -gy + margin, options, 0, (LPCWSTR) &glyph, 1, 0);
     }
     
     SelectObject(hdc, old_font);
@@ -1237,7 +1417,7 @@
     if (qt_cleartype_enabled) {
         LOGFONT lf = logfont;
         lf.lfQuality = ANTIALIASED_QUALITY;
-        font = CreateFontIndirect(&lf);
+        font = CreateFontIndirectW(&lf);
     }
     QImage::Format mask_format = QNativeImage::systemFormat();
 #ifndef Q_OS_WINCE
@@ -1362,9 +1542,17 @@
     QString fam = fallbacks.at(at-1);
 
     LOGFONT lf = static_cast<QFontEngineWin *>(engines.at(0))->logfont;
-    memcpy(lf.lfFaceName, fam.utf16(), sizeof(wchar_t) * qMin(fam.length() + 1, 32));  // 32 = Windows hard-coded
-    HFONT hfont = CreateFontIndirect(&lf);
-
+    HFONT hfont;
+    QT_WA({
+        memcpy(lf.lfFaceName, fam.utf16(), sizeof(TCHAR)*qMin(fam.length()+1,32));  // 32 = Windows hard-coded
+        hfont = CreateFontIndirectW(&lf);
+    } , {
+        // LOGFONTA and LOGFONTW are binary compatible
+        QByteArray lname = fam.toLocal8Bit();
+        memcpy(lf.lfFaceName,lname.data(),
+            qMin(lname.length()+1,32));  // 32 = Windows hard-coded
+        hfont = CreateFontIndirectA((LOGFONTA*)&lf);
+    });
     bool stockFont = false;
     if (hfont == 0) {
         hfont = (HFONT)GetStockObject(ANSI_VAR_FONT);
